Replace legacy pulse drawer

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
close #5392 

This PR deprecates current (legacy) pulse drawer and upgrades `.draw` method of `Schedule` and `Pulse` to use v2 drawer.

### Details and comments

Currently v2 drawer is placed under `qiskit.visualization.pulse_v2` and legacy drawer is under `qiskit.visualization.pulse`.
I plan to replace the drawer with following steps.

In the next release:
- Deprecate legacy drawer and associated stylesheets.
- Release v2 drawer.
- Update `.draw` methods.

3 months after:
- Remove legacy drawer and associated stylesheets.
- Move `qiskit.visualization.pulse_v2` to `qiskit.visualization.pulse` with symbolic link in `pulse_v2`.
- Deprecate import path `qiskit.visualization.pulse_v2`

6 months after:
- Remove `qiskit.visualization.pulse_v2`

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
-             interp_method: Optional[Callable] = None,
-             scale: float = 1, interactive: bool = False,
-             draw_title: bool = False):
+    def draw(self,
```
Order of arguments should be maintained for backward compatibility where code input kwargs as args. As a sidenote this is why everyone should use `kwarg=kwarg` when calling a function with a kwarg, to prevent rearrangement of kwargs from breaking code.

After deprecation, these can be dropped.

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+                :py:class:~`qiskit.visualization.pulse_v2.stylesheets.IQXSimple`, and
+                :py:class:~`qiskit.visualization.pulse_v2.stylesheets.IQXDebugging` for details of
+                preset stylesheets.
+            backend: Backend object to play the input pulse program. If this object is provided,
+                the input program is visualized with the details of hardware information.
```
```suggestion
            backend: Backend object to play the input pulse program. If provided, the plotter
                may use to make the visualization hardware aware.
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+                preset stylesheets.
+            backend: Backend object to play the input pulse program. If this object is provided,
+                the input program is visualized with the details of hardware information.
+            time_range: Set horizontal axis limit. Tuple `(tmin, tmax)`.
```
```suggestion
            time_range: Set horizontal axis limit. Tuple ``(tmin, tmax)``.
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+            backend: Backend object to play the input pulse program. If this object is provided,
+                the input program is visualized with the details of hardware information.
+            time_range: Set horizontal axis limit. Tuple `(tmin, tmax)`.
+            time_unit: The unit of specified time range either `dt` or `ns`.
+                The unit of `ns` is available only when `backend` object is provided.
```
```suggestion
            time_unit: The unit of specified time range either ``dt`` or ``ns``.
                The unit of ``ns`` is available only when ``backend`` object is provided.
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+            time_unit: The unit of specified time range either `dt` or `ns`.
+                The unit of `ns` is available only when `backend` object is provided.
+            show_waveform_info: A control property to show annotations, i.e. name, of waveforms.
+                Set `True` to show additional information about waveforms.
```
```suggestion
                Set ``True`` to show additional information about waveforms.
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+            time_range: Set horizontal axis limit. Tuple `(tmin, tmax)`.
+            time_unit: The unit of specified time range either `dt` or `ns`.
+                The unit of `ns` is available only when `backend` object is provided.
+            show_waveform_info: A control property to show annotations, i.e. name, of waveforms.
```
```suggestion
            show_waveform_info: Show waveform annotations, i.e. name, of waveforms.
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+        # pylint: disable=invalid-name, cyclic-import, missing-return-type-doc
+        from qiskit.visualization import pulse_drawer_v2
+
+        legacy_args = (dt, filename, interp_method, scale, interactive, draw_title)
+
+        if any(arg is not None for arg in legacy_args):
+            warnings.warn('Legacy pulse drawer is deprecated with some arguments. '
+                          'Please check the API document of new pulse drawer '
+                          '`qiskit.visualization.pulse_drawer_v2`.',
+                          DeprecationWarning)
```
Should warn which argument is deprecated.

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -59,29 +61,85 @@ def assign_parameters(self,
+                          DeprecationWarning)
+
+        if filename:
+            warnings.warn('File saving is delegated to the plotter software in new drawer. '
+                          'If you specify matplotlib plotter family to `plotter` argument, '
+                          'you can call `savefig` method with the returned Figure object.',
+                          DeprecationWarning)
```
```suggestion
            warnings.warn('File saving is delegated to the plotter software in the new drawer. '
                          'If a matplotlib plotter family is passed to the "plotter" argument, '
                          'the "savefig" method may be called with the returned Figure object.',
                          DeprecationWarning)
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+            time_unit: The unit of specified time range either `dt` or `ns`.
+                The unit of `ns` is available only when `backend` object is provided.
+            disable_channels: A control property to show specific pulse channel.
+                Pulse channel instances provided as a list is not shown in the output image.
```
```suggestion
                Pulse channel instances provided as a list are not shown in the output image.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                The unit of `ns` is available only when `backend` object is provided.
+            disable_channels: A control property to show specific pulse channel.
+                Pulse channel instances provided as a list is not shown in the output image.
+            show_snapshot: A control property to show snapshot symbols.
+                Set `True` to show snapshot instructions.
```
```suggestion
            show_snapshot: Show snapshot instructions.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                Pulse channel instances provided as a list is not shown in the output image.
+            show_snapshot: A control property to show snapshot symbols.
+                Set `True` to show snapshot instructions.
+            show_framechange: A control property to show frame change symbols.
+                Set `True` to show frame change instructions. The frame change represents
+                instructions that modulate phase or frequency of pulse channels.
```
```suggestion
            show_framechange: Show frame change instructions. The frame change represents
                instructions that modulate phase or frequency of pulse channels.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+            show_framechange: A control property to show frame change symbols.
+                Set `True` to show frame change instructions. The frame change represents
+                instructions that modulate phase or frequency of pulse channels.
+            show_waveform_info: A control property to show annotations, i.e. name, of waveforms.
+                Set `True` to show additional information about waveforms.
```
```suggestion
            show_waveform_info: Show additional information about waveforms such as their name.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                instructions that modulate phase or frequency of pulse channels.
+            show_waveform_info: A control property to show annotations, i.e. name, of waveforms.
+                Set `True` to show additional information about waveforms.
+            show_barrier: A control property to show barriers.
+                Set `True` to show barrier lines.
```
```suggestion
            show_barrier: Show barrier lines.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+
+                    mpl2d: Matplotlib API for 2D image generation.
+                        Matplotlib API to generate 2D image. Charts are placed along y axis with
+                        vertical offset. This API takes matplotlib.axes.Axes as `axis` input.
```
```suggestion
                        vertical offset. This API takes matplotlib.axes.Axes as ``axis`` input.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                        Matplotlib API to generate 2D image. Charts are placed along y axis with
+                        vertical offset. This API takes matplotlib.axes.Axes as `axis` input.
+
+                `axis` and `style` kwargs may depend on the plotter.
```
```suggestion
                ``axis`` and ``style`` kwargs may depend on the plotter.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+
+                `axis` and `style` kwargs may depend on the plotter.
+            axis: Arbitrary object passed to the plotter. If this object is provided,
+                the plotters uses given `axis` instead of internally initializing a figure object.
```
```suggestion
                the plotters use a given ``axis`` instead of internally initializing a figure object.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                This object format depends on the plotter. See plotters section for details.
+            dt: Deprecated. This argument is used by the legacy pulse drawer.
+            filename: Deprecated. This argument is used by the legacy pulse drawer.
+                To save output image, you can call `.savefig` method with
```
```suggestion
                To save output image, you can call ``.savefig`` method with
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
-            matplotlib.Figure: A matplotlib figure object of the pulse schedule.
+            Visualization output data.
+
+            The returned data type depends on the `plotter`.
```
```suggestion
            The returned data type depends on the ``plotter``.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+            Visualization output data.
+
+            The returned data type depends on the `plotter`.
+            If matplotlib family is specified, this will be a `matplotlib.pyplot.Figure` data.
```
```suggestion
            If matplotlib family is specified, this will be a ``matplotlib.pyplot.Figure`` data.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+
+            The returned data type depends on the `plotter`.
+            If matplotlib family is specified, this will be a `matplotlib.pyplot.Figure` data.
+            The returned data is generated by the `.get_image` method of the specified plotter API.
```
```suggestion
            The returned data is generated by the ``.get_image`` method of the specified plotter API.
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+        # pylint: disable=invalid-name, cyclic-import, missing-return-type-doc
+        from qiskit.visualization import pulse_drawer_v2
+
+        legacy_args = (dt, filename, interp_method, scale, channel_scales, plot_all,
+                       plot_range, interactive, table, label, framechange, channels,
+                       show_framechange_channels, draw_title)
+
+        if any(arg is not None for arg in legacy_args):
+            warnings.warn('Legacy pulse drawer is deprecated with some arguments. '
+                          'Please check the API document of new pulse drawer '
+                          '`qiskit.visualization.pulse_drawer_v2`.',
+                          DeprecationWarning)
```
Should report what is deprecated for convenience.
```suggestion
        legacy_args = (dt, filename, interp_method, scale, channel_scales, plot_all,
                       plot_range, interactive, table, label, framechange, channels,
                       show_framechange_channels, draw_title)

        if any(arg is not None for arg in legacy_args):
            warnings.warn('Legacy pulse drawer is deprecated with some arguments. '
                          'Please check the API document of new pulse drawer '
                          '`qiskit.visualization.pulse_drawer_v2`.',
                          DeprecationWarning)
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -773,63 +773,122 @@ def _update_parameter_table(self, schedule: 'Schedule'):
+                          '`qiskit.visualization.pulse_drawer_v2`.',
+                          DeprecationWarning)
+
+        if filename:
+            warnings.warn('File saving is delegated to the plotter software in new drawer. '
+                          'If you specify matplotlib plotter family to `plotter` argument, '
+                          'you can call `savefig` method with the returned Figure object.',
+                          DeprecationWarning)
```
Update as above.

taalexander commented on releasenotes/notes/replace-pulse-drawer-f9f667c8f71e1e02.yaml:
```diff
@@ -0,0 +1,25 @@
+---
+upgrade:
+  - |
+    New pulse drawer :func:`qiskit.visualization.pulse_drawer_v2` is released.
```
```suggestion
    The new pulse drawer :func:`qiskit.visualization.pulse_drawer_v2` is now released.
```

taalexander commented on releasenotes/notes/replace-pulse-drawer-f9f667c8f71e1e02.yaml:
```diff
@@ -0,0 +1,25 @@
+    New pulse drawer :func:`qiskit.visualization.pulse_drawer_v2` is released.
+    This drawer supports multiple new features:
+
+    * Truncation of long pulse instructions
```
```suggestion
    * Truncation of long pulse instructions.
```

taalexander commented on releasenotes/notes/replace-pulse-drawer-f9f667c8f71e1e02.yaml:
```diff
@@ -0,0 +1,25 @@
+
+    * Truncation of long pulse instructions
+    * Visualization of parametric pulses.
+    * New stylesheets ``IQXStandard``, ``IQXSimple``, ``IQXDebugging``
```
```suggestion
    * New stylesheets ``IQXStandard``, ``IQXSimple``, ``IQXDebugging``.
```

taalexander commented on releasenotes/notes/replace-pulse-drawer-f9f667c8f71e1e02.yaml:
```diff
@@ -0,0 +1,25 @@
+    * Truncation of long pulse instructions
+    * Visualization of parametric pulses.
+    * New stylesheets ``IQXStandard``, ``IQXSimple``, ``IQXDebugging``
+    * Visualization of system info (channel frequency, etc...) by specifying ``backend`` object.
```
```suggestion
    * Visualization of system info (channel frequency, etc...) by specifying ``backend`` objects for visualization.
```

taalexander commented on releasenotes/notes/replace-pulse-drawer-f9f667c8f71e1e02.yaml:
```diff
@@ -0,0 +1,25 @@
+    * Visualization of parametric pulses.
+    * New stylesheets ``IQXStandard``, ``IQXSimple``, ``IQXDebugging``
+    * Visualization of system info (channel frequency, etc...) by specifying ``backend`` object.
+    * Specifying ``axis`` object to create complex image data, i.e. for publication.
```
```suggestion
    * Specifying ``axis`` objects for plotting to allow further extension of generated plots, i.e., for publication manipulations.
```

taalexander reviewed (changes requested):
Looks great, excited to see this go out. Have the old plotter tests been validated against these changes https://github.com/Qiskit/qiskit-terra/blob/master/test/python/visualization/test_pulse_visualization_output.py?

nkanazawa1989 commented:
Thanks @taalexander for reviewing! PR is updated (mainly to rearrange argument ordering and docstrings). I also fixed unittests of legacy drawer to explicitly call `pulse_drawer` function rather than draw method. 

Along with these changes I also updated the behavior of debugging stylesheet (apply pulse phase modulation by default). Previously this was just disabled and it shows waveform as-is and we can see the frame with framechange info. Based on my usability test experience, this was confusing because shape of pulse depends on stylesheet. Now all stylesheets apply phase modulation by default (user can disable phase modulation by initializing a stylesheet with `formatter.control.apply_phase_modulation: False`.

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -132,6 +146,11 @@ def draw(self,
 
+        if isinstance(style, PulseStyle):
+            style = None
+            warnings.warn('Legacy stylesheet is specified. This is just ignored in new drawer. '
```
```suggestion
            warnings.warn('Legacy stylesheet is specified. This is ignored in the new drawer. '
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -113,17 +112,32 @@ def draw(self,
+
+        if active_legacy_args:
+            warnings.warn('Legacy pulse drawer is deprecated. '
+                          'Specified arguments {dep_args} are all deprecated too. '
```
```suggestion
                          'Specified arguments {dep_args} are deprecated. '
```

taalexander commented on qiskit/pulse/library/pulse.py:
```diff
@@ -132,6 +146,11 @@ def draw(self,
+        if isinstance(style, PulseStyle):
+            style = None
+            warnings.warn('Legacy stylesheet is specified. This is just ignored in new drawer. '
+                          'Please check API document of this method.')
```
```suggestion
                          'Please check the API documentation for this method.')
```

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -854,21 +850,38 @@ def draw(self,
+
+        if active_legacy_args:
+            warnings.warn('Legacy pulse drawer is deprecated. '
+                          'Specified arguments {dep_args} are all deprecated too. '
```
```suggestion
                          'Specified arguments {dep_args} are deprecated. '
```

taalexander commented on qiskit/visualization/pulse_v2/interface.py:
````diff
@@ -355,17 +353,17 @@ def draw(program: Union[Waveform, ParametricPulse, Schedule],
-                'formatter.channel_scaling.measure': 5
-            }
-            style = IQXStandard(**my_style)
+            ```python
````
This is markdown style syntax, please use a [code-block](https://www.sphinx-doc.org/en/1.5/markup/code.html#directive-code-block) instead.

taalexander reviewed (changes requested):
Just a couple of small updates left. The docs are also still failing to build.
